File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.java
class Solution {
    private boolean dfs(int node, int destination, Map<Integer, List<Integer>> graph, Set<Integer> visited) {
        if (node == destination) {
            return true;
        }
        
        visited.add(node);
        
        for (int neighbour : graph.get(node)) {
            if (!visited.contains(neighbour)) {
                if (dfs(neighbour, destination, graph, visited)) {
                    return true;
                }
            }
        }
        
        return false;
    }
    
    public boolean validPath(int n, int[][] edges, int source, int destination) {
        Map<Integer, List<Integer>> graph = new HashMap();
        
        for (int i = 0; i < n; i++) {
            graph.put(i, new ArrayList<>());
        }
        
        for (int[] edge : edges) {
            int node1 = edge[0];
            int node2 = edge[1];
            graph.get(node1).add(node2);
            graph.get(node2).add(node1);
        }
        
        Set<Integer> visited = new HashSet();
        return dfs(source, destination, graph, visited);
    }
}

File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.py
from collections import defaultdict
from collections import deque

class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        graph = defaultdict(list)
        for a, b in edges:
            graph[a].append(b)
            graph[b].append(a)
        
        visited = set()
        queue = deque([source])
        
        while queue:
            currentNode = queue.popleft()   #edges = [[0,1],[1,2],[2,0]],source=0, destination=2
            if currentNode == destination:
                return True
            for nextNode in graph[currentNode]:
                if nextNode not in visited:
                    visited.add(nextNode)
                    queue.append(nextNode)
        
        return False
